Cilium Daemon set crash on "panic: runtime error: invalid memory address or nil pointer dereference" after few requests

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Implmented Cilium as a full kubeProxy & CNI Replacement on EKS cluster. 
Aiming to work with Gateway API CRDs.

While performing several successful HTTP requests to backend service , cilium Daemon set is crashing after few requests(one digit count):
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5fb794]

goroutine 1630 [running]:
net/url.(*URL).String(0x0)
	/usr/local/go/src/net/url/url.go:810 +0x34
github.com/cilium/cilium/pkg/hubble/parser/seven.decodeHTTP({0x3928cce, 0x7}, 0xc003903680, 0xc00029c660)
	/go/src/github.com/cilium/cilium/pkg/hubble/parser/seven/http.go:41 +0x405
github.com/cilium/cilium/pkg/hubble/parser/seven.decodeLayer7(0x192?, 0xc0037f9bc0?)
	/go/src/github.com/cilium/cilium/pkg/hubble/parser/seven/parser.go:360 +0x2b9
github.com/cilium/cilium/pkg/hubble/parser/seven.(*Parser).Decode(0xc00024de30, 0xc0021bf6c0, 0xc0038ff380)
	/go/src/github.com/cilium/cilium/pkg/hubble/parser/seven/parser.go:149 +0xbc5
github.com/cilium/cilium/pkg/hubble/parser.(*Parser).Decode(0xc00029c740, 0xc003903810)
	/go/src/github.com/cilium/cilium/pkg/hubble/parser/parser.go:145 +0x2e5
github.com/cilium/cilium/pkg/hubble/observer.(*LocalObserverServer).Start(0xc0008528c0)
	/go/src/github.com/cilium/cilium/pkg/hubble/observer/local_observer.go:133 +0xe6
created by github.com/cilium/cilium/daemon/cmd.(*Daemon).launchHubble in goroutine 1532
	/go/src/github.com/cilium/cilium/daemon/cmd/hubble.go:211 +0x1bdb
```

Cilium installed with the following options:
```
      eni.enabled: true 
      ipam.mode: eni 
      kubeProxyReplacement: true
      nodePort.enabled: true
      egressMasqueradeInterfaces: eth0 
      routingMode: native
      k8sServiceHost: <KUBEAPI SERVER>
      k8sServicePort: 443
      auto-create-cilium-node-resource: "true"
      gatewayAPI.enabled: true 
      hubble.enabled: true
      hubble.relay.enabled: true 
      hubble.ui.enabled: true
      envoy.enabled: true
```

HTTP Route
```
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name:  <name>
spec:
  hostnames:
  - <hostname>
  parentRefs:
  - name: stm-gateway
    namespace: stm
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /monitoring/v1/test
    filters:
      - type: RequestHeaderModifier
        requestHeaderModifier:
          add:
            - name: X-Forwarded-Proto
              value: https
      - type: URLRewrite
        urlRewrite:
          path:
            type: ReplacePrefixMatch
            replacePrefixMatch: /
    backendRefs:
    - name: <service_name>
      port: 80
```
Gateway resource:
```
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: test-gateway
spec:
  gatewayClassName: cilium
  infrastructure: 
    annotations:
      service.kubernetes.io/topology-aware-hints: Auto
      service.kubernetes.io/topology-aware-routing: Auto
      service.beta.kubernetes.io/aws-load-balancer-additional-resource-tags: businessunit=shared,system=kubernetes
      service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
      service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
      service.beta.kubernetes.io/aws-load-balancer-scheme: internal
      service.beta.kubernetes.io/aws-load-balancer-ssl-cert: <CERT_ARN>
      service.beta.kubernetes.io/aws-load-balancer-ssl-ports: port-443
  listeners:
  - protocol: HTTP
    port: 80
    name: http
    allowedRoutes:
       namespaces:
        from: Same
    hostname: <hostname>
  - protocol: HTTP
    port: 443
    name: https
    allowedRoutes:
       namespaces:
        from: Same
    hostname: <hostname>
```

### Cilium Version

cilium image (stable): v1.15.1
cilium image (running): 1.15.1

### Kernel Version

5.10.198-187.748.amzn2.x86_64

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.5-eks-5e0fdde", GitCommit:"e78a4be9da4c375a87a109e0f4a5f4a8d2bc17c0", GitTreeState:"clean", BuildDate:"2024-01-02T20:34:46Z", GoVersion:"go1.20.12", Compiler:"gc", Platform:"linux/amd64"}

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/hubble/parser/seven/http.go
+++ b/pkg/hubble/parser/seven/http.go
@@ -5,7 +5,7 @@ package seven
 
 import (
 	"fmt"
-	neturl "net/url"
+	"net/url"
 	"sort"
 	"strings"
 
@@ -29,8 +29,8 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 			headers = append(headers, &flowpb.HTTPHeader{Key: key, Value: filteredValue})
 		}
 	}
-	url, _ := neturl.Parse(http.URL.String())
-	filterURL(url, opts.HubbleRedactSettings)
+	uri, _ := url.Parse(http.URL.String())
+	filterURL(uri, opts.HubbleRedactSettings)
 
 	if flowType == accesslog.TypeRequest {
 		// Set only fields that are relevant for requests.
@@ -38,7 +38,7 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 			Http: &flowpb.HTTP{
 				Method:   http.Method,
 				Protocol: http.Protocol,
-				Url:      url.String(),
+				Url:      uri.String(),
 				Headers:  headers,
 			},
 		}
@@ -49,16 +49,16 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 			Code:     uint32(http.Code),
 			Method:   http.Method,
 			Protocol: http.Protocol,
-			Url:      url.String(),
+			Url:      uri.String(),
 			Headers:  headers,
 		},
 	}
 }
 
 func (p *Parser) httpSummary(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, flow *flowpb.Flow) string {
-	url, _ := neturl.Parse(http.URL.String())
-	filterURL(url, p.opts.HubbleRedactSettings)
-	httpRequest := http.Method + " " + url.String()
+	uri, _ := url.Parse(http.URL.String())
+	filterURL(uri, p.opts.HubbleRedactSettings)
+	httpRequest := http.Method + " " + uri.String()
 	switch flowType {
 	case accesslog.TypeRequest:
 		return fmt.Sprintf("%s %s", http.Protocol, httpRequest)
@@ -105,16 +105,16 @@ func filterHeader(key string, value string, redactSettings options.HubbleRedactS
 // conditionally redacts sensitive values.
 // If configured and user info exists, it removes the password from the flow.
 // If configured, it removes the URL's query parts from the flow.
-func filterURL(url *neturl.URL, redactSettings options.HubbleRedactSettings) {
-	if url != nil {
-		if redactSettings.RedactHTTPUserInfo && url.User != nil {
-			if _, ok := url.User.Password(); ok {
-				url.User = neturl.UserPassword(url.User.Username(), defaults.SensitiveValueRedacted)
+func filterURL(uri *url.URL, redactSettings options.HubbleRedactSettings) {
+	if uri != nil {
+		if redactSettings.RedactHTTPUserInfo && uri.User != nil {
+			if _, ok := uri.User.Password(); ok {
+				uri.User = url.UserPassword(uri.User.Username(), defaults.SensitiveValueRedacted)
 			}
 		}
 		if redactSettings.RedactHTTPQuery {
-			url.RawQuery = ""
-			url.Fragment = ""
+			uri.RawQuery = ""
+			uri.Fragment = ""
 		}
 	}
 }

--- a/pkg/hubble/parser/seven/http.go
+++ b/pkg/hubble/parser/seven/http.go
@@ -29,7 +29,7 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 			headers = append(headers, &flowpb.HTTPHeader{Key: key, Value: filteredValue})
 		}
 	}
-	uri, _ := url.Parse(http.URL.String())
+	uri := cloneURL(http.URL)
 	filterURL(uri, opts.HubbleRedactSettings)
 
 	if flowType == accesslog.TypeRequest {
@@ -56,7 +56,7 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 }
 
 func (p *Parser) httpSummary(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, flow *flowpb.Flow) string {
-	uri, _ := url.Parse(http.URL.String())
+	uri := cloneURL(http.URL)
 	filterURL(uri, p.opts.HubbleRedactSettings)
 	httpRequest := http.Method + " " + uri.String()
 	switch flowType {
@@ -117,4 +117,18 @@ func filterURL(uri *url.URL, redactSettings options.HubbleRedactSettings) {
 			uri.Fragment = ""
 		}
 	}
+}
+
+// cloneURL return a copy of the given URL. Copied from src/net/http/clone.go.
+func cloneURL(u *url.URL) *url.URL {
+	if u == nil {
+		return nil
+	}
+	u2 := new(url.URL)
+	*u2 = *u
+	if u.User != nil {
+		u2.User = new(url.Userinfo)
+		*u2.User = *u.User
+	}
+	return u2
 }

--- a/pkg/hubble/parser/seven/http.go
+++ b/pkg/hubble/parser/seven/http.go
@@ -29,8 +29,7 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 			headers = append(headers, &flowpb.HTTPHeader{Key: key, Value: filteredValue})
 		}
 	}
-	uri := cloneURL(http.URL)
-	filterURL(uri, opts.HubbleRedactSettings)
+	uri := filteredURL(http.URL, opts.HubbleRedactSettings)
 
 	if flowType == accesslog.TypeRequest {
 		// Set only fields that are relevant for requests.
@@ -56,8 +55,7 @@ func decodeHTTP(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, opts
 }
 
 func (p *Parser) httpSummary(flowType accesslog.FlowType, http *accesslog.LogRecordHTTP, flow *flowpb.Flow) string {
-	uri := cloneURL(http.URL)
-	filterURL(uri, p.opts.HubbleRedactSettings)
+	uri := filteredURL(http.URL, p.opts.HubbleRedactSettings)
 	httpRequest := http.Method + " " + uri.String()
 	switch flowType {
 	case accesslog.TypeRequest:
@@ -101,22 +99,27 @@ func filterHeader(key string, value string, redactSettings options.HubbleRedactS
 	return value
 }
 
-// filterURL receives the observed URL and a set of Hubble redact settings and
-// conditionally redacts sensitive values.
+// filteredURL return a copy of the given URL potentially mutated depending on
+// Hubble redact settings.
 // If configured and user info exists, it removes the password from the flow.
 // If configured, it removes the URL's query parts from the flow.
-func filterURL(uri *url.URL, redactSettings options.HubbleRedactSettings) {
-	if uri != nil {
-		if redactSettings.RedactHTTPUserInfo && uri.User != nil {
-			if _, ok := uri.User.Password(); ok {
-				uri.User = url.UserPassword(uri.User.Username(), defaults.SensitiveValueRedacted)
-			}
-		}
-		if redactSettings.RedactHTTPQuery {
-			uri.RawQuery = ""
-			uri.Fragment = ""
+func filteredURL(uri *url.URL, redactSettings options.HubbleRedactSettings) *url.URL {
+	if uri == nil {
+		// NOTE: return a non-nil URL so that we can always call String() on
+		// it.
+		return &url.URL{}
+	}
+	u2 := cloneURL(uri)
+	if redactSettings.RedactHTTPUserInfo && u2.User != nil {
+		if _, ok := u2.User.Password(); ok {
+			u2.User = url.UserPassword(u2.User.Username(), defaults.SensitiveValueRedacted)
 		}
 	}
+	if redactSettings.RedactHTTPQuery {
+		u2.RawQuery = ""
+		u2.Fragment = ""
+	}
+	return u2
 }
 
 // cloneURL return a copy of the given URL. Copied from src/net/http/clone.go.

--- a/pkg/hubble/parser/seven/http_test.go
+++ b/pkg/hubble/parser/seven/http_test.go
@@ -378,6 +378,61 @@ func TestGetL7HTTPResponseTraceID(t *testing.T) {
 	assert.False(t, ok, "request id should not be in the cache")
 }
 
+// see https://github.com/cilium/cilium/issues/31071
+func TestDecodeL7HTTPWithInvalidURL(t *testing.T) {
+	requestPath, err := url.Parse("http://myhost/some/path")
+	require.NoError(t, err)
+
+	// mutate requestPath such as url.Parse(requestPath.String()) fails, which
+	// triggered the panic described in #31071.
+	requestPath.Host += "@" // invalid hostname
+	_, err = url.Parse(requestPath.String())
+	require.Error(t, err)
+
+	lr := &accesslog.LogRecord{
+		Type:                accesslog.TypeRequest,
+		Timestamp:           fakeTimestamp,
+		NodeAddressInfo:     fakeNodeInfo,
+		ObservationPoint:    accesslog.Ingress,
+		SourceEndpoint:      fakeSourceEndpoint,
+		DestinationEndpoint: fakeDestinationEndpoint,
+		IPVersion:           accesslog.VersionIPv4,
+		Verdict:             accesslog.VerdictForwarded,
+		TransportProtocol:   accesslog.TransportProtocol(u8proto.TCP),
+		ServiceInfo:         nil,
+		DropReason:          nil,
+		HTTP: &accesslog.LogRecordHTTP{
+			Code:     0,
+			Method:   "POST",
+			URL:      requestPath,
+			Protocol: "HTTP/1.1",
+			Headers: http.Header{
+				"Host":        {"myhost"},
+				"Traceparent": {"00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+			},
+		},
+	}
+
+	parser, err := New(log, nil, nil, nil, nil)
+	require.NoError(t, err)
+
+	f := &flowpb.Flow{}
+	err = parser.Decode(lr, f)
+	require.NoError(t, err)
+
+	assert.Equal(t, &flowpb.HTTP{
+		Code:     0,
+		Method:   "POST",
+		Url:      "http://myhost%40/some/path",
+		Protocol: "HTTP/1.1",
+		Headers: []*flowpb.HTTPHeader{
+			{Key: "Host", Value: "myhost"},
+			{Key: "Traceparent", Value: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+		},
+	}, f.GetL7().GetHttp())
+	assert.Equal(t, "4bf92f3577b34da6a3ce929d0e0e4736", f.GetTraceContext().GetParent().GetTraceId())
+}
+
 func TestDecodeL7HTTPRequestRemoveUrlQuery(t *testing.T) {
 	requestPath, err := url.Parse("http://myhost/some/path?foo=bar")
 	require.NoError(t, err)

--- a/pkg/hubble/parser/seven/http_test.go
+++ b/pkg/hubble/parser/seven/http_test.go
@@ -433,6 +433,51 @@ func TestDecodeL7HTTPWithInvalidURL(t *testing.T) {
 	assert.Equal(t, "4bf92f3577b34da6a3ce929d0e0e4736", f.GetTraceContext().GetParent().GetTraceId())
 }
 
+func TestDecodeL7HTTPWithNilURL(t *testing.T) {
+	lr := &accesslog.LogRecord{
+		Type:                accesslog.TypeRequest,
+		Timestamp:           fakeTimestamp,
+		NodeAddressInfo:     fakeNodeInfo,
+		ObservationPoint:    accesslog.Ingress,
+		SourceEndpoint:      fakeSourceEndpoint,
+		DestinationEndpoint: fakeDestinationEndpoint,
+		IPVersion:           accesslog.VersionIPv4,
+		Verdict:             accesslog.VerdictForwarded,
+		TransportProtocol:   accesslog.TransportProtocol(u8proto.TCP),
+		ServiceInfo:         nil,
+		DropReason:          nil,
+		HTTP: &accesslog.LogRecordHTTP{
+			Code:     0,
+			Method:   "POST",
+			URL:      nil,
+			Protocol: "HTTP/1.1",
+			Headers: http.Header{
+				"Host":        {"myhost"},
+				"Traceparent": {"00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+			},
+		},
+	}
+
+	parser, err := New(log, nil, nil, nil, nil)
+	require.NoError(t, err)
+
+	f := &flowpb.Flow{}
+	err = parser.Decode(lr, f)
+	require.NoError(t, err)
+
+	assert.Equal(t, &flowpb.HTTP{
+		Code:     0,
+		Method:   "POST",
+		Url:      "",
+		Protocol: "HTTP/1.1",
+		Headers: []*flowpb.HTTPHeader{
+			{Key: "Host", Value: "myhost"},
+			{Key: "Traceparent", Value: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+		},
+	}, f.GetL7().GetHttp())
+	assert.Equal(t, "4bf92f3577b34da6a3ce929d0e0e4736", f.GetTraceContext().GetParent().GetTraceId())
+}
+
 func TestDecodeL7HTTPRequestRemoveUrlQuery(t *testing.T) {
 	requestPath, err := url.Parse("http://myhost/some/path?foo=bar")
 	require.NoError(t, err)

--- a/pkg/hubble/parser/seven/http_test.go
+++ b/pkg/hubble/parser/seven/http_test.go
@@ -143,54 +143,6 @@ func TestDecodeL7HTTPRequest(t *testing.T) {
 	assert.Equal(t, "4bf92f3577b34da6a3ce929d0e0e4736", f.GetTraceContext().GetParent().GetTraceId())
 }
 
-func TestDecodeL7HTTPRequestRemoveUrlQuery(t *testing.T) {
-	requestPath, err := url.Parse("http://myhost/some/path?foo=bar")
-	require.NoError(t, err)
-	lr := &accesslog.LogRecord{
-		Type:                accesslog.TypeRequest,
-		Timestamp:           fakeTimestamp,
-		NodeAddressInfo:     fakeNodeInfo,
-		ObservationPoint:    accesslog.Ingress,
-		SourceEndpoint:      fakeSourceEndpoint,
-		DestinationEndpoint: fakeDestinationEndpoint,
-		IPVersion:           accesslog.VersionIPv4,
-		Verdict:             accesslog.VerdictForwarded,
-		TransportProtocol:   accesslog.TransportProtocol(u8proto.TCP),
-		ServiceInfo:         nil,
-		DropReason:          nil,
-		HTTP: &accesslog.LogRecordHTTP{
-			Code:     0,
-			Method:   "POST",
-			URL:      requestPath,
-			Protocol: "HTTP/1.1",
-			Headers: http.Header{
-				"Host":        {"myhost"},
-				"Traceparent": {"00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
-			},
-		},
-	}
-	lr.SourceEndpoint.Port = 56789
-	lr.DestinationEndpoint.Port = 80
-
-	opts := []options.Option{options.Redact(nil, true, true, false, []string{}, []string{"authorization"})}
-	parser, err := New(log, nil, nil, nil, nil, opts...)
-	require.NoError(t, err)
-
-	f := &flowpb.Flow{}
-	err = parser.Decode(lr, f)
-	require.NoError(t, err)
-	assert.Equal(t, &flowpb.HTTP{
-		Code:     0,
-		Method:   "POST",
-		Url:      "http://myhost/some/path",
-		Protocol: "HTTP/1.1",
-		Headers: []*flowpb.HTTPHeader{
-			{Key: "Host", Value: "myhost"},
-			{Key: "Traceparent", Value: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
-		},
-	}, f.GetL7().GetHttp())
-}
-
 func TestDecodeL7HTTPRecordResponse(t *testing.T) {
 	requestPath, err := url.Parse("http://myhost/some/path")
 	require.NoError(t, err)
@@ -424,6 +376,54 @@ func TestGetL7HTTPResponseTraceID(t *testing.T) {
 	assert.Empty(t, f.GetTraceContext().GetParent().GetTraceId())
 	_, ok = parser.traceContextCache.Get(requestID)
 	assert.False(t, ok, "request id should not be in the cache")
+}
+
+func TestDecodeL7HTTPRequestRemoveUrlQuery(t *testing.T) {
+	requestPath, err := url.Parse("http://myhost/some/path?foo=bar")
+	require.NoError(t, err)
+	lr := &accesslog.LogRecord{
+		Type:                accesslog.TypeRequest,
+		Timestamp:           fakeTimestamp,
+		NodeAddressInfo:     fakeNodeInfo,
+		ObservationPoint:    accesslog.Ingress,
+		SourceEndpoint:      fakeSourceEndpoint,
+		DestinationEndpoint: fakeDestinationEndpoint,
+		IPVersion:           accesslog.VersionIPv4,
+		Verdict:             accesslog.VerdictForwarded,
+		TransportProtocol:   accesslog.TransportProtocol(u8proto.TCP),
+		ServiceInfo:         nil,
+		DropReason:          nil,
+		HTTP: &accesslog.LogRecordHTTP{
+			Code:     0,
+			Method:   "POST",
+			URL:      requestPath,
+			Protocol: "HTTP/1.1",
+			Headers: http.Header{
+				"Host":        {"myhost"},
+				"Traceparent": {"00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+			},
+		},
+	}
+	lr.SourceEndpoint.Port = 56789
+	lr.DestinationEndpoint.Port = 80
+
+	opts := []options.Option{options.Redact(nil, true, true, false, []string{}, []string{"authorization"})}
+	parser, err := New(log, nil, nil, nil, nil, opts...)
+	require.NoError(t, err)
+
+	f := &flowpb.Flow{}
+	err = parser.Decode(lr, f)
+	require.NoError(t, err)
+	assert.Equal(t, &flowpb.HTTP{
+		Code:     0,
+		Method:   "POST",
+		Url:      "http://myhost/some/path",
+		Protocol: "HTTP/1.1",
+		Headers: []*flowpb.HTTPHeader{
+			{Key: "Host", Value: "myhost"},
+			{Key: "Traceparent", Value: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"},
+		},
+	}, f.GetL7().GetHttp())
 }
 
 func TestDecodeL7HTTPRequestHeadersRedact(t *testing.T) {

